<a href="https://colab.research.google.com/github/LatiefDataVisionary/data-science-capstone-project-college/blob/main/notebooks/02_eda_raw_data.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Exploratory Data Analysis of Raw Spotify Reviews**

This notebook performs an initial Exploratory Data Analysis (EDA) on the raw dataset of Spotify app reviews. The goal is to understand the basic structure, quality, and characteristics of the data before any preprocessing steps are applied.

## **1. Setup and Library Imports**

In [45]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from wordcloud import WordCloud

# Set plot style
sns.set_style('whitegrid')

## **2. Data Loading and Initial Inspection**

Loading the dataset from the specified CSV file into a pandas DataFrame.

In [46]:
# English comment dataset
data_en_1 = 'https://raw.githubusercontent.com/LatiefDataVisionary/data-science-capstone-project-college/refs/heads/main/data/raw/spotify_reviews_en_au_1.csv'
data_en_2 = 'https://raw.githubusercontent.com/LatiefDataVisionary/data-science-capstone-project-college/refs/heads/main/data/raw/spotify_reviews_en_au_2.csv'
data_en_3 = 'https://raw.githubusercontent.com/LatiefDataVisionary/data-science-capstone-project-college/refs/heads/main/data/raw/spotify_reviews_en_uk_1.csv'
data_en_4 = 'https://raw.githubusercontent.com/LatiefDataVisionary/data-science-capstone-project-college/refs/heads/main/data/raw/spotify_reviews_en_uk_2.csv'
data_en_5 = 'https://raw.githubusercontent.com/LatiefDataVisionary/data-science-capstone-project-college/refs/heads/main/data/raw/spotify_reviews_en_us_1.csv'
data_en_6 = 'https://raw.githubusercontent.com/LatiefDataVisionary/data-science-capstone-project-college/refs/heads/main/data/raw/spotify_reviews_en_us_2.csv'

# Indonesian comment dataset
data_id_1 = 'https://raw.githubusercontent.com/LatiefDataVisionary/data-science-capstone-project-college/refs/heads/main/data/raw/spotify_reviews_id_id_1.csv'
data_id_2 = 'https://raw.githubusercontent.com/LatiefDataVisionary/data-science-capstone-project-college/refs/heads/main/data/raw/spotify_reviews_id_id_2.csv'

In [47]:
# Create empty lists to store dataframes
en_dfs = []
id_dfs = []

In [48]:
# List of English data file paths
english_files = [data_en_1, data_en_2, data_en_3, data_en_4, data_en_5, data_en_6]

# Iterate through English files and load them
for file_path in english_files:
    try:
        df = pd.read_csv(file_path)
        en_dfs.append(df)
        print(f"Successfully loaded {file_path}")
    except FileNotFoundError:
        print(f"Error: {file_path} not found.")
    except Exception as e:
        print(f"Error loading {file_path}: {e}")


# Concatenate English dataframes
if en_dfs:
    raw_en_df = pd.concat(en_dfs, ignore_index=True)
    print(f"\nEnglish dataset loaded successfully with {len(raw_en_df)} reviews.")
else:
    raw_en_df = None
    print("\nNo English dataframes were loaded.")

Successfully loaded https://raw.githubusercontent.com/LatiefDataVisionary/data-science-capstone-project-college/refs/heads/main/data/raw/spotify_reviews_en_au_1.csv
Successfully loaded https://raw.githubusercontent.com/LatiefDataVisionary/data-science-capstone-project-college/refs/heads/main/data/raw/spotify_reviews_en_au_2.csv
Successfully loaded https://raw.githubusercontent.com/LatiefDataVisionary/data-science-capstone-project-college/refs/heads/main/data/raw/spotify_reviews_en_uk_1.csv
Successfully loaded https://raw.githubusercontent.com/LatiefDataVisionary/data-science-capstone-project-college/refs/heads/main/data/raw/spotify_reviews_en_uk_2.csv
Successfully loaded https://raw.githubusercontent.com/LatiefDataVisionary/data-science-capstone-project-college/refs/heads/main/data/raw/spotify_reviews_en_us_1.csv
Successfully loaded https://raw.githubusercontent.com/LatiefDataVisionary/data-science-capstone-project-college/refs/heads/main/data/raw/spotify_reviews_en_us_2.csv

English d

In [49]:
# List of Indonesian data file paths
indonesian_files = [data_id_1, data_id_2]

# Iterate through Indonesian files and load them
for file_path in indonesian_files:
    try:
        df = pd.read_csv(file_path)
        id_dfs.append(df)
        print(f"Successfully loaded {file_path}")
    except FileNotFoundError:
        print(f"Error: {file_path} not found.")
    except Exception as e:
        print(f"Error loading {file_path}: {e}")


# Concatenate Indonesian dataframes
if id_dfs:
    raw_id_df = pd.concat(id_dfs, ignore_index=True)
    print(f"\nIndonesian dataset loaded successfully with {len(raw_id_df)} reviews.")
else:
    raw_id_df = None
    print("\nNo Indonesian dataframes were loaded.")

Successfully loaded https://raw.githubusercontent.com/LatiefDataVisionary/data-science-capstone-project-college/refs/heads/main/data/raw/spotify_reviews_id_id_1.csv
Successfully loaded https://raw.githubusercontent.com/LatiefDataVisionary/data-science-capstone-project-college/refs/heads/main/data/raw/spotify_reviews_id_id_2.csv

Indonesian dataset loaded successfully with 65000 reviews.


Displaying the first 5 rows of the DataFrame to get a glimpse of the data structure and content.

In [50]:
if raw_en_df is not None:
    print("First 5 rows of English Reviews DataFrame (raw_en_df):")
    display(raw_en_df.head())

First 5 rows of English Reviews DataFrame (raw_en_df):


reviewId       userName  \
0  97b3a330-8135-4961-963b-d5b40aeaa80a  A Google user   
1  39a62a75-3998-483e-aca2-8719d3f8dd57  A Google user   
2  03c513ab-df54-42a2-aa3c-7502c086d875  A Google user   
3  5f769b79-7d95-4129-8078-1dc62d8f2b2b  A Google user   
4  fd4665c6-ed09-4af0-94a2-06b8f5e2d8c4  A Google user   

                                           userImage  \
0  https://play-lh.googleusercontent.com/EGemoI2N...   
1  https://play-lh.googleusercontent.com/EGemoI2N...   
2  https://play-lh.googleusercontent.com/EGemoI2N...   
3  https://play-lh.googleusercontent.com/EGemoI2N...   
4  https://play-lh.googleusercontent.com/EGemoI2N...   

                                             content  score  thumbsUpCount  \
0  they fixed it, I was just really pissy yesterd...      5              1   
1  Offline doesnt work, support doesnt help, just...      1              0   
2  Super annoying ad experience! It feels like th...      1              5   
3                                                  👍      5              1   
4                          super song for everything      5              0   

  reviewCreatedVersion                   at  \
0          9.0.82.1032  2025-09-27 22:44:08   
1          9.0.82.1032  2025-09-27 21:48:01   
2           9.0.56.591  2025-09-27 06:05:50   
3          9.0.82.1032  2025-09-27 06:02:10   
4          9.0.80.1343  2025-09-27 06:01:26   

                                        replyContent            repliedAt  \
0                                                NaN                  NaN   
1  Hello. Thanks for bringing this to our attenti...  2025-09-27 21:52:49   
2                                                NaN                  NaN   
3                                                NaN                  NaN   
4                                                NaN                  NaN   

    appVersion  
0  9.0.82.1032  
1  9.0.82.1032  
2   9.0.56.591  
3  9.0.82.1032  
4  9.0.80.1343

In [51]:
if raw_id_df is not None:
    print("\nFirst 5 rows of Indonesian Reviews DataFrame (raw_id_df):")
    display(raw_id_df.head())


First 5 rows of Indonesian Reviews DataFrame (raw_id_df):


reviewId         userName  \
0  427e2299-b27a-4f53-89ac-15f9885207c8  Pengguna Google   
1  790cee66-b937-41ba-b7a8-e831c60f63a4  Pengguna Google   
2  2a59f6d0-fa47-4e26-b097-9ef0ddc824f6  Pengguna Google   
3  0d356a00-bd64-4539-a23f-aaa7e46c98d8  Pengguna Google   
4  0349c341-fad0-4e60-8952-28dc1cb43659  Pengguna Google   

                                           userImage  \
0  https://play-lh.googleusercontent.com/EGemoI2N...   
1  https://play-lh.googleusercontent.com/EGemoI2N...   
2  https://play-lh.googleusercontent.com/EGemoI2N...   
3  https://play-lh.googleusercontent.com/EGemoI2N...   
4  https://play-lh.googleusercontent.com/EGemoI2N...   

                                             content  score  thumbsUpCount  \
0  lagu bukan hanya alunan nada tapi bisa jadi un...      1              2   
1  iklan Mulu gak jelass apa apa harus premium ko...      1              0   
2                           Terima kasih banyak 🙏👍👍👍      5              0   
3  kok di aku mah gk bisa ada lirik ya sih tolong...      4              0   
4                             sangat banyak lagu nya      5              0   

  reviewCreatedVersion                   at replyContent repliedAt  \
0                  NaN  2025-09-28 05:29:11          NaN       NaN   
1          9.0.82.1032  2025-09-28 03:08:45          NaN       NaN   
2                  NaN  2025-09-27 06:05:37          NaN       NaN   
3          9.0.82.1032  2025-09-27 06:03:26          NaN       NaN   
4          9.0.80.1343  2025-09-27 05:56:16          NaN       NaN   

    appVersion  
0          NaN  
1  9.0.82.1032  
2          NaN  
3  9.0.82.1032  
4  9.0.80.1343

Printing concise information about the DataFrame, including the index dtype and column dtypes, non-null values, and memory usage. This helps in understanding the data types and identifying potential missing values.

In [52]:
if raw_id_df is not None:
    print("\nThe informaton of English Reviews DataFrame (raw_enn_df):")
    display(raw_en_df.info())


The informaton of English Reviews DataFrame (raw_enn_df):
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 155000 entries, 0 to 154999
Data columns (total 11 columns):
 #   Column                Non-Null Count   Dtype 
---  ------                --------------   ----- 
 0   reviewId              155000 non-null  object
 1   userName              155000 non-null  object
 2   userImage             155000 non-null  object
 3   content               154992 non-null  object
 4   score                 155000 non-null  int64 
 5   thumbsUpCount         155000 non-null  int64 
 6   reviewCreatedVersion  134926 non-null  object
 7   at                    155000 non-null  object
 8   replyContent          14061 non-null   object
 9   repliedAt             14061 non-null   object
 10  appVersion            134926 non-null  object
dtypes: int64(2), object(9)
memory usage: 13.0+ MB


None

In [53]:
if raw_id_df is not None:
    print("\nThe informaiton of Indonesian Reviews DataFrame (raw_id_df):")
    display(raw_id_df.info())


The informaiton of Indonesian Reviews DataFrame (raw_id_df):
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65000 entries, 0 to 64999
Data columns (total 11 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   reviewId              65000 non-null  object
 1   userName              65000 non-null  object
 2   userImage             65000 non-null  object
 3   content               65000 non-null  object
 4   score                 65000 non-null  int64 
 5   thumbsUpCount         65000 non-null  int64 
 6   reviewCreatedVersion  48732 non-null  object
 7   at                    65000 non-null  object
 8   replyContent          1439 non-null   object
 9   repliedAt             1439 non-null   object
 10  appVersion            48732 non-null  object
dtypes: int64(2), object(9)
memory usage: 5.5+ MB


None

Generating descriptive statistics for both the numerical and categorical columns in the DataFrame. This provides insights into the central tendency, dispersion, and shape of the distribution of numerical data.

In [54]:
if raw_en_df is not None:
    print("Descriptive statistics for English Reviews DataFrame (raw_en_df):")
    display(raw_en_df.describe(include='all'))

Descriptive statistics for English Reviews DataFrame (raw_en_df):


reviewId       userName  \
count                                 155000         155000   
unique                                 74069           2994   
top     d72295a1-8f55-49ed-b629-1c191de4ef62  A Google user   
freq                                       6         151993   
mean                                     NaN            NaN   
std                                      NaN            NaN   
min                                      NaN            NaN   
25%                                      NaN            NaN   
50%                                      NaN            NaN   
75%                                      NaN            NaN   
max                                      NaN            NaN   

                                                userImage content  \
count                                              155000  154992   
unique                                               3003   65943   
top     https://play-lh.googleusercontent.com/EGemoI2N...    good   
freq                                               151993    2234   
mean                                                  NaN     NaN   
std                                                   NaN     NaN   
min                                                   NaN     NaN   
25%                                                   NaN     NaN   
50%                                                   NaN     NaN   
75%                                                   NaN     NaN   
max                                                   NaN     NaN   

                score  thumbsUpCount reviewCreatedVersion  \
count   155000.000000  155000.000000               134926   
unique            NaN            NaN                  498   
top               NaN            NaN           9.0.68.632   
freq              NaN            NaN                18752   
mean         4.206206       8.154148                  NaN   
std          1.461646     173.687554                  NaN   
min          1.000000       0.000000                  NaN   
25%          4.000000       0.000000                  NaN   
50%          5.000000       0.000000                  NaN   
75%          5.000000       0.000000                  NaN   
max          5.000000   20643.000000                  NaN   

                         at  \
count                155000   
unique                73891   
top     2025-09-16 12:24:03   
freq                     12   
mean                    NaN   
std                     NaN   
min                     NaN   
25%                     NaN   
50%                     NaN   
75%                     NaN   
max                     NaN   

                                             replyContent  \
count                                               14061   
unique                                               2252   
top     Hey! We hear you and we'll pass your feedback ...   
freq                                                 1216   
mean                                                  NaN   
std                                                   NaN   
min                                                   NaN   
25%                                                   NaN   
50%                                                   NaN   
75%                                                   NaN   
max                                                   NaN   

                  repliedAt  appVersion  
count                 14061      134926  
unique                 5802         498  
top     2025-09-07 15:52:39  9.0.68.632  
freq                      6       18752  
mean                    NaN         NaN  
std                     NaN         NaN  
min                     NaN         NaN  
25%                     NaN         NaN  
50%                     NaN         NaN  
75%                     NaN         NaN  
max                     NaN         NaN

In [55]:
if raw_id_df is not None:
    print("\nDescriptive statistics for Indonesian Reviews DataFrame (raw_id_df):")
    display(raw_id_df.describe(include='all'))


Descriptive statistics for Indonesian Reviews DataFrame (raw_id_df):


reviewId         userName  \
count                                  65000            65000   
unique                                 59183             1614   
top     1c362127-ea44-4322-b1de-436f45904463  Pengguna Google   
freq                                       2            63385   
mean                                     NaN              NaN   
std                                      NaN              NaN   
min                                      NaN              NaN   
25%                                      NaN              NaN   
50%                                      NaN              NaN   
75%                                      NaN              NaN   
max                                      NaN              NaN   

                                                userImage content  \
count                                               65000   65000   
unique                                               1615   44694   
top     https://play-lh.googleusercontent.com/EGemoI2N...   bagus   
freq                                                63385    1833   
mean                                                  NaN     NaN   
std                                                   NaN     NaN   
min                                                   NaN     NaN   
25%                                                   NaN     NaN   
50%                                                   NaN     NaN   
75%                                                   NaN     NaN   
max                                                   NaN     NaN   

               score  thumbsUpCount reviewCreatedVersion                   at  \
count   65000.000000   65000.000000                48732                65000   
unique           NaN            NaN                  411                59111   
top              NaN            NaN            9.0.2.459  2025-09-25 14:14:58   
freq             NaN            NaN                 3276                    4   
mean        4.337231       5.492569                  NaN                  NaN   
std         1.340434      78.316078                  NaN                  NaN   
min         1.000000       0.000000                  NaN                  NaN   
25%         5.000000       0.000000                  NaN                  NaN   
50%         5.000000       0.000000                  NaN                  NaN   
75%         5.000000       0.000000                  NaN                  NaN   
max         5.000000    7893.000000                  NaN                  NaN   

                                             replyContent  \
count                                                1439   
unique                                                563   
top     Hai, apakah Anda sudah mencoba menginstal ulan...   
freq                                                   90   
mean                                                  NaN   
std                                                   NaN   
min                                                   NaN   
25%                                                   NaN   
50%                                                   NaN   
75%                                                   NaN   
max                                                   NaN   

                  repliedAt appVersion  
count                  1439      48732  
unique                 1434        411  
top     2025-08-14 15:58:40  9.0.2.459  
freq                      2       3276  
mean                    NaN        NaN  
std                     NaN        NaN  
min                     NaN        NaN  
25%                     NaN        NaN  
50%                     NaN        NaN  
75%                     NaN        NaN  
max                     NaN        NaN

## **3. Data Quality Check**

Checking for missing values in each column and summing them up to get a total count of missing entries per column.

In [56]:
if raw_en_df is not None:
    print("Missing values per column (English dataset - raw_en_df):")
    missing_values_en = raw_en_df.isnull().sum()
    print(missing_values_en)

Missing values per column (English dataset - raw_en_df):
reviewId                     0
userName                     0
userImage                    0
content                      8
score                        0
thumbsUpCount                0
reviewCreatedVersion     20074
at                           0
replyContent            140939
repliedAt               140939
appVersion               20074
dtype: int64


In [57]:
if raw_id_df is not None:
    print("\nMissing values per column (Indonesian dataset - raw_id_df):")
    missing_values_id = raw_id_df.isnull().sum()
    print(missing_values_id)


Missing values per column (Indonesian dataset - raw_id_df):
reviewId                    0
userName                    0
userImage                   0
content                     0
score                       0
thumbsUpCount               0
reviewCreatedVersion    16268
at                          0
replyContent            63561
repliedAt               63561
appVersion              16268
dtype: int64


Checking for and counting duplicate rows in the DataFrame. Duplicate rows can skew analysis and should be identified.

In [65]:
if raw_en_df is not None:
    duplicate_rows_en = raw_en_df.duplicated().sum()
    print(f"\nNumber of duplicate rows (English dataset - raw_en_df): {duplicate_rows_en}")
    print(f'The total length: {len(raw_en_df)}')


Number of duplicate rows (English dataset - raw_en_df): 77922
The total length: 155000


In [64]:
if raw_id_df is not None:
    duplicate_rows_id = raw_id_df.duplicated().sum()
    print(f"Number of duplicate rows (Indonesian dataset - raw_id_df): {duplicate_rows_id}")
    print(f'The total length: {len(raw_id_df)}')

Number of duplicate rows (Indonesian dataset - raw_id_df): 5754
The total length: 65000


**Summary of Data Quality:**

Based on the checks above:
- We can see the number of missing values in each column.
- We have identified the total count of duplicate rows.

This initial assessment helps us understand the cleanliness of the raw data and highlights areas that may require attention during preprocessing.

## 4. Analysis of Review Ratings (score)

Visualizing the distribution of review scores using a count plot to understand the frequency of each rating (1 to 5 stars).

In [ ]:
if raw_df is not None:
    plt.figure(figsize=(8, 6))
    sns.countplot(data=raw_df, x='score', palette='viridis')
    plt.title('Distribution of Review Scores')
    plt.xlabel('Score')
    plt.ylabel('Number of Reviews')
    plt.show()

**Analysis of Score Distribution:**

The count plot reveals the distribution of star ratings. We can observe which scores are most frequent. This gives us an initial idea of the overall sentiment of the reviews and the balance of the dataset. For example, if one score significantly outweighs others, the dataset is skewed.

## 5. Analysis of Review Text (content)

### Subsection 5.1: Review Length Analysis

Creating new columns for character length and word count of each review in the `content` column.

In [ ]:
if raw_df is not None and 'content' in raw_df.columns:
    raw_df['char_length'] = raw_df['content'].astype(str).apply(len)
    raw_df['word_count'] = raw_df['content'].astype(str).apply(lambda x: len(x.split()))
    print("Character length and word count columns added.")

Generating histograms to visualize the distribution of review lengths based on character count and word count.

In [ ]:
if raw_df is not None and 'char_length' in raw_df.columns and 'word_count' in raw_df.columns:
    fig, axes = plt.subplots(1, 2, figsize=(14, 6))

    sns.histplot(data=raw_df, x='char_length', bins=50, ax=axes[0], kde=True)
    axes[0].set_title('Distribution of Review Character Length')
    axes[0].set_xlabel('Character Length')
    axes[0].set_ylabel('Frequency')

    sns.histplot(data=raw_df, x='word_count', bins=50, ax=axes[1], kde=True)
    axes[1].set_title('Distribution of Review Word Count')
    axes[1].set_xlabel('Word Count')
    axes[1].set_ylabel('Frequency')

    plt.tight_layout()
    plt.show()

**Analysis of Review Length:**

The histograms show the distribution of how long reviews are, both in terms of the number of characters and the number of words. This helps us understand the typical length of reviews and identify any outliers (very short or very long reviews). Most reviews appear to be concentrated within a certain range of words/characters.

### Subsection 5.2: Word Cloud of Raw Text

Generating a Word Cloud from the raw review text to visually represent the most frequent words. Note that this is based on unprocessed text and will include stopwords and other noise.

In [ ]:
if raw_df is not None and 'content' in raw_df.columns:
    # Combine all reviews into a single string
    all_reviews = " ".join(review for review in raw_df['content'].astype(str))

    # Generate word cloud
    wordcloud = WordCloud(width=800, height=400, background_color='white').generate(all_reviews)

    # Display the word cloud
    plt.figure(figsize=(10, 7))
    plt.imshow(wordcloud, interpolation='bilinear')
    plt.axis("off")
    plt.title('Word Cloud of Raw Review Text')
    plt.show()

**Analysis of Word Cloud:**

The word cloud provides a quick visual overview of the most frequently occurring words in the raw review text. As expected with unprocessed text, it contains many common words (stopwords) and potential noise. However, it gives a raw first impression of the dominant themes and keywords present in the reviews before any cleaning or preprocessing.

## 6. Temporal Analysis (Review Timestamps)

Converting the `at` column to a datetime object to enable time-based analysis. Errors in conversion will be coerced to `NaT` (Not a Time).

In [ ]:
if raw_df is not None and 'at' in raw_df.columns:
    raw_df['at'] = pd.to_datetime(raw_df['at'], errors='coerce')
    print(" 'at' column converted to datetime objects.")
    # Drop rows where datetime conversion failed
    raw_df.dropna(subset=['at'], inplace=True)
    print(f"Dropped rows with invalid datetime values. Remaining rows: {len(raw_df)}")

Plotting the number of reviews submitted over time, grouped by month, to identify any trends or patterns in review activity.

In [ ]:
if raw_df is not None and 'at' in raw_df.columns:
    # Extract month and year
    raw_df['review_month'] = raw_df['at'].dt.to_period('M')

    # Group by month and count reviews
    monthly_reviews = raw_df.groupby('review_month').size()

    # Plot the time series
    plt.figure(figsize=(12, 6))
    monthly_reviews.plot()
    plt.title('Number of Reviews Over Time (Monthly)')
    plt.xlabel('Date')
    plt.ylabel('Number of Reviews')
    plt.xticks(rotation=45)
    plt.tight_layout()
    plt.show()

**Analysis of Temporal Trends:**

The time series plot shows how the number of reviews has changed over the months. We can observe trends such as periods of increased or decreased review activity. Spikes in reviews might correlate with significant events like app updates, marketing campaigns, or service outages. A general upward or downward trend could indicate changes in user engagement over time.

## 7. Conclusion & Next Steps

**Conclusion:**

This initial EDA on the raw Spotify reviews dataset has provided valuable insights into its structure, quality, and content. We have examined:
- The dataset's basic information and structure.
- The presence of missing values and duplicate rows.
- The distribution of review scores, indicating the overall sentiment.
- The characteristics of the review text, including length and the most frequent raw words.
- The temporal trends in review submission.

The raw text data, as seen in the word cloud, is noisy and requires significant cleaning and preprocessing before it can be effectively used for tasks like sentiment analysis or topic modeling.

**Next Steps:**

The logical next step in this project, which will be covered in the notebook `03_data_preprocessing_and_cleaning.ipynb`, is to clean and preprocess the raw text data. This will involve steps such as:
- Handling missing values and duplicate rows (if not already done).
- Text cleaning (e.g., removing punctuation, special characters, numbers).
- Lowercasing the text.
- Tokenization.
- Removing stopwords.
- Stemming or lemmatization.
- Handling bilingual text (if necessary).

These preprocessing steps are crucial to transform the raw review text into a format suitable for further analysis and machine learning models.

# Task
Generate the complete code and markdown text for a Jupyter Notebook named 02_eda_raw_data.ipynb to perform a comprehensive Exploratory Data Analysis (EDA) on raw, unprocessed bilingual (English and Indonesian) Spotify app reviews. The dataset consists of multiple CSV files for each language located in "../data/raw/". The notebook should include sections for setup, data loading and initial inspection, data quality check, analysis of review ratings, analysis of review text (including length and word clouds), temporal analysis, and a conclusion summarizing findings and outlining next steps. The entire output, including titles, explanations, variable names, and comments, must be in English. The analysis should be performed for both English and Indonesian reviews, using separate dataframes (`raw_en_df` and `raw_id_df`) where appropriate, and the code should handle potential errors during data loading and processing.

## Modify data loading

### Subtask:
Update the data loading cell to load the English and Indonesian datasets into `raw_en_df` and `raw_id_df` DataFrames respectively, concatenating the multiple files for each language. Handle potential errors during loading.


## Modify initial inspection

### Subtask:
Update the initial inspection cells (`.head()`, `.info()`, `.describe()`) to display information for both `raw_en_df` and `raw_id_df`.


## Modify data quality check

### Subtask:
Update the data quality check cells (missing values and duplicates) to perform checks on both `raw_en_df` and `raw_id_df`.


**Reasoning**:
Update the code to check for missing values and duplicate rows in both `raw_en_df` and `raw_id_df` and print the results with clear labels.



In [ ]:
if raw_en_df is not None:
    print("Missing values per column (English dataset - raw_en_df):")
    missing_values_en = raw_en_df.isnull().sum()
    print(missing_values_en)

if raw_id_df is not None:
    print("\nMissing values per column (Indonesian dataset - raw_id_df):")
    missing_values_id = raw_id_df.isnull().sum()
    print(missing_values_id)

if raw_en_df is not None:
    duplicate_rows_en = raw_en_df.duplicated().sum()
    print(f"\nNumber of duplicate rows (English dataset - raw_en_df): {duplicate_rows_en}")

if raw_id_df is not None:
    duplicate_rows_id = raw_id_df.duplicated().sum()
    print(f"Number of duplicate rows (Indonesian dataset - raw_id_df): {duplicate_rows_id}")

## Modify analysis of review ratings

### Subtask:
Update the score distribution analysis and plotting to visualize the distribution for both English and Indonesian reviews, potentially on separate plots or a combined plot with language as a differentiator.


**Reasoning**:
Update the score distribution analysis by combining the English and Indonesian dataframes and plotting the score distribution for both languages on the same plot using 'language' as a hue.



In [ ]:
if raw_en_df is not None and raw_id_df is not None and 'score' in raw_en_df.columns and 'score' in raw_id_df.columns:
    # Add a language column to differentiate between English and Indonesian reviews
    raw_en_df['language'] = 'English'
    raw_id_df['language'] = 'Indonesian'

    # Combine the dataframes
    combined_df = pd.concat([raw_en_df, raw_id_df], ignore_index=True)

    # Create the count plot
    plt.figure(figsize=(10, 6))
    sns.countplot(data=combined_df, x='score', hue='language', palette='viridis')
    plt.title('Distribution of Review Scores by Language')
    plt.xlabel('Score')
    plt.ylabel('Number of Reviews')
    plt.show()
elif raw_en_df is not None and 'score' in raw_en_df.columns:
    print("Only English data available. Plotting English score distribution.")
    plt.figure(figsize=(8, 6))
    sns.countplot(data=raw_en_df, x='score', palette='viridis')
    plt.title('Distribution of English Review Scores')
    plt.xlabel('Score')
    plt.ylabel('Number of Reviews')
    plt.show()
elif raw_id_df is not None and 'score' in raw_id_df.columns:
    print("Only Indonesian data available. Plotting Indonesian score distribution.")
    plt.figure(figsize=(8, 6))
    sns.countplot(data=raw_id_df, x='score', palette='viridis')
    plt.title('Distribution of Indonesian Review Scores')
    plt.xlabel('Score')
    plt.ylabel('Number of Reviews')
    plt.show()
else:
    print("Score data is not available in either dataframe for plotting.")

## Modify analysis of review text

### Subtask:
Update the analysis of review text to include review length analysis (character length and word count) and word cloud generation for both English and Indonesian reviews.


**Reasoning**:
Calculate character length and word count for both English and Indonesian dataframes.



In [ ]:
if raw_en_df is not None and 'content' in raw_en_df.columns:
    raw_en_df['char_length'] = raw_en_df['content'].astype(str).apply(len)
    raw_en_df['word_count'] = raw_en_df['content'].astype(str).apply(lambda x: len(x.split()))
    print("Character length and word count columns added to raw_en_df.")
else:
    print("raw_en_df is None or 'content' column not found in raw_en_df. Cannot add length/count columns.")

if raw_id_df is not None and 'content' in raw_id_df.columns:
    raw_id_df['char_length'] = raw_id_df['content'].astype(str).apply(len)
    raw_id_df['word_count'] = raw_id_df['content'].astype(str).apply(lambda x: len(x.split()))
    print("Character length and word count columns added to raw_id_df.")
else:
    print("raw_id_df is None or 'content' column not found in raw_id_df. Cannot add length/count columns.")

**Reasoning**:
Generate histograms to visualize the distribution of character length and word count for both English and Indonesian reviews using subplots.



In [ ]:
if raw_en_df is not None and 'char_length' in raw_en_df.columns and 'word_count' in raw_en_df.columns and \
   raw_id_df is not None and 'char_length' in raw_id_df.columns and 'word_count' in raw_id_df.columns:

    fig, axes = plt.subplots(2, 2, figsize=(16, 12))

    sns.histplot(data=raw_en_df, x='char_length', bins=50, ax=axes[0, 0], kde=True, color='skyblue')
    axes[0, 0].set_title('Distribution of English Review Character Length')
    axes[0, 0].set_xlabel('Character Length')
    axes[0, 0].set_ylabel('Frequency')

    sns.histplot(data=raw_en_df, x='word_count', bins=50, ax=axes[0, 1], kde=True, color='salmon')
    axes[0, 1].set_title('Distribution of English Review Word Count')
    axes[0, 1].set_xlabel('Word Count')
    axes[0, 1].set_ylabel('Frequency')

    sns.histplot(data=raw_id_df, x='char_length', bins=50, ax=axes[1, 0], kde=True, color='skyblue')
    axes[1, 0].set_title('Distribution of Indonesian Review Character Length')
    axes[1, 0].set_xlabel('Character Length')
    axes[1, 0].set_ylabel('Frequency')

    sns.histplot(data=raw_id_df, x='word_count', bins=50, ax=axes[1, 1], kde=True, color='salmon')
    axes[1, 1].set_title('Distribution of Indonesian Review Word Count')
    axes[1, 1].set_xlabel('Word Count')
    axes[1, 1].set_ylabel('Frequency')

    plt.tight_layout()
    plt.show()

elif raw_en_df is not None and 'char_length' in raw_en_df.columns and 'word_count' in raw_en_df.columns:
    print("Only English data available. Plotting English review length distributions.")
    fig, axes = plt.subplots(1, 2, figsize=(14, 6))
    sns.histplot(data=raw_en_df, x='char_length', bins=50, ax=axes[0], kde=True, color='skyblue')
    axes[0].set_title('Distribution of English Review Character Length')
    axes[0].set_xlabel('Character Length')
    axes[0].set_ylabel('Frequency')
    sns.histplot(data=raw_en_df, x='word_count', bins=50, ax=axes[1], kde=True, color='salmon')
    axes[1].set_title('Distribution of English Review Word Count')
    axes[1].set_xlabel('Word Count')
    axes[1].set_ylabel('Frequency')
    plt.tight_layout()
    plt.show()

elif raw_id_df is not None and 'char_length' in raw_id_df.columns and 'word_count' in raw_id_df.columns:
    print("Only Indonesian data available. Plotting Indonesian review length distributions.")
    fig, axes = plt.subplots(1, 2, figsize=(14, 6))
    sns.histplot(data=raw_id_df, x='char_length', bins=50, ax=axes[0], kde=True, color='skyblue')
    axes[0].set_title('Distribution of Indonesian Review Character Length')
    axes[0].set_xlabel('Character Length')
    axes[0].set_ylabel('Frequency')
    sns.histplot(data=raw_id_df, x='word_count', bins=50, ax=axes[1], kde=True, color='salmon')
    axes[1].set_title('Distribution of Indonesian Review Word Count')
    axes[1].set_xlabel('Word Count')
    axes[1].set_ylabel('Frequency')
    plt.tight_layout()
    plt.show()

else:
    print("Review length data is not available in either dataframe for plotting.")

**Reasoning**:
Generate and display separate word clouds for the raw review content of English and Indonesian dataframes.



In [ ]:
if raw_en_df is not None and 'content' in raw_en_df.columns:
    # Combine all English reviews into a single string
    all_en_reviews = " ".join(review for review in raw_en_df['content'].astype(str))

    # Generate English word cloud
    wordcloud_en = WordCloud(width=800, height=400, background_color='white').generate(all_en_reviews)

    # Display the English word cloud
    plt.figure(figsize=(10, 7))
    plt.imshow(wordcloud_en, interpolation='bilinear')
    plt.axis("off")
    plt.title('Word Cloud of Raw English Review Text')
    plt.show()
else:
    print("raw_en_df is None or 'content' column not found in raw_en_df. Cannot generate English word cloud.")


if raw_id_df is not None and 'content' in raw_id_df.columns:
    # Combine all Indonesian reviews into a single string
    all_id_reviews = " ".join(review for review in raw_id_df['content'].astype(str))

    # Generate Indonesian word cloud
    wordcloud_id = WordCloud(width=800, height=400, background_color='white').generate(all_id_reviews)

    # Display the Indonesian word cloud
    plt.figure(figsize=(10, 7))
    plt.imshow(wordcloud_id, interpolation='bilinear')
    plt.axis("off")
    plt.title('Word Cloud of Raw Indonesian Review Text')
    plt.show()
else:
    print("raw_id_df is None or 'content' column not found in raw_id_df. Cannot generate Indonesian word cloud.")

## Modify temporal analysis

### Subtask:
Update the temporal analysis to plot the number of reviews over time for both English and Indonesian datasets, potentially on the same plot with different lines/colors or on separate subplots.


**Reasoning**:
Check for dataframe existence and 'at' column, convert 'at' to datetime, drop NaT rows, extract month, group by month and count reviews for both dataframes.



In [ ]:
if raw_en_df is not None and 'at' in raw_en_df.columns:
    raw_en_df['at'] = pd.to_datetime(raw_en_df['at'], errors='coerce')
    raw_en_df.dropna(subset=['at'], inplace=True)
    raw_en_df['review_month'] = raw_en_df['at'].dt.to_period('M')
    monthly_reviews_en = raw_en_df.groupby('review_month').size()
    print(f"English reviews processed. Remaining rows: {len(raw_en_df)}")
else:
    monthly_reviews_en = None
    print("raw_en_df is None or 'at' column not found in raw_en_df. Cannot process English temporal data.")

if raw_id_df is not None and 'at' in raw_id_df.columns:
    raw_id_df['at'] = pd.to_datetime(raw_id_df['at'], errors='coerce')
    raw_id_df.dropna(subset=['at'], inplace=True)
    raw_id_df['review_month'] = raw_id_df['at'].dt.to_period('M')
    monthly_reviews_id = raw_id_df.groupby('review_month').size()
    print(f"Indonesian reviews processed. Remaining rows: {len(raw_id_df)}")
else:
    monthly_reviews_id = None
    print("raw_id_df is None or 'at' column not found in raw_id_df. Cannot process Indonesian temporal data.")

**Reasoning**:
Plot the monthly review counts for both English and Indonesian reviews on the same plot, adding title, labels, legend, and rotating x-axis ticks.



In [ ]:
plt.figure(figsize=(12, 6))

if monthly_reviews_en is not None:
    monthly_reviews_en.plot(label='English Reviews')

if monthly_reviews_id is not None:
    monthly_reviews_id.plot(label='Indonesian Reviews')

if monthly_reviews_en is not None or monthly_reviews_id is not None:
    plt.title('Number of Reviews Over Time (Monthly) by Language')
    plt.xlabel('Date')
    plt.ylabel('Number of Reviews')
    plt.xticks(rotation=45)
    plt.legend()
    plt.tight_layout()
    plt.show()
else:
    print("No temporal review data available for plotting.")

## Modify conclusion

### Subtask:
Update the conclusion to summarize the findings for both English and Indonesian datasets and reiterate the next steps for cleaning and preprocessing, acknowledging the bilingual nature of the data.


**Reasoning**:
Update the conclusion markdown cell to summarize the findings for both English and Indonesian datasets and reiterate the next steps, acknowledging the bilingual nature of the data.



In [ ]:
# This is a markdown cell, the content will be updated directly in the notebook.
# No Python code is needed here.

## Summary:

### Data Analysis Key Findings

*   The English dataset contains 155,000 reviews and the Indonesian dataset contains 65,000 reviews. Both datasets have 11 columns.
*   Both English and Indonesian datasets have missing values in the `content`, `reviewCreatedVersion`, `replyContent`, `repliedAt`, and `appVersion` columns, with varying extents of missingness between the languages.
*   Significant numbers of duplicate rows were found in both datasets: 77922 in the English dataset and 5754 in the Indonesian dataset.
*   The average review score is slightly higher for Indonesian reviews (4.34) compared to English reviews (4.21).
*   The distribution of `thumbsUpCount` is highly skewed in both languages, with most reviews having zero likes.
*   Review length analysis showed the distributions of character length and word count for both English and Indonesian reviews.
*   Temporal analysis revealed the monthly trend of review submissions for both languages.

### Insights or Next Steps

*   Address the significant number of duplicate rows in both datasets during the data cleaning phase.
*   Develop a strategy to handle missing values, particularly in columns like `replyContent` and `repliedAt`, which have a high percentage of missing data.
